packages/@aws-cdk/aws-codepipeline-api/lib/action.ts

```diff
@@ -0,0 +1,266 @@
+            detailType: [ 'CodePipeline Stage Execution State Change' ],
+            source: [ 'aws.codepipeline' ],
+            resources: [ this.stage.pipelineArn ],
+            detail: {
+                stage: [ this.stage.name ],
+                action: [ this.name ],
+            },
+        });
+        return rule;
+    }
+
+    protected addChild(child: cdk.Construct, name: string) {
+        super.addChild(child, name);
+        if (child instanceof Artifact) {
+            this.outputArtifacts.push(child);
+        }
+    }
+
+    protected addOutputArtifact(name: string): Artifact {
+        const artifact = new Artifact(this, name);
+        return artifact;
+    }
+
+    protected addInputArtifact(artifact: Artifact): Action {
+        this.inputArtifacts.push(artifact);
+        return this;
+    }
+}
+
+// export class TestAction extends Action {
+//     constructor(parent: Stage, name: string, provider: string, artifactBounds: ActionArtifactBounds, configuration?: any) {
+//         super(parent, name, {
+//           category: ActionCategory.Test,
+//           provider,
+//           artifactBounds,
+//           configuration
+//         });
+//     }
+// }
+
+// export class CodeBuildTest extends TestAction {
+//     constructor(parent: Stage, name: string, project: codebuild.ProjectArnAttribute) {
+//         super(parent, name, 'CodeBuild', { minInputs: 1, maxInputs: 1, minOutputs: 0, maxOutputs: 1 }, {
+//             ProjectName: project
+//         });
+//     }
+// }
+
+// export class CodeDeploy extends DeployAction {
+//     constructor(parent: Stage, name: string, applicationName: string, deploymentGroupName: string) {
+//         super(parent, name, 'CodeDeploy', { minInputs: 1, maxInputs: 1, minOutputs: 0, maxOutputs: 0 }, {
+//             ApplicationName: applicationName,
+//             DeploymentGroupName: deploymentGroupName
+//         });
+//     }
+// }
+
+// export class ElasticBeanstalkDeploy extends DeployAction {
+//     constructor(parent: Stage, name: string, applicationName: string, environmentName: string) {
+//         super(parent, name, 'ElasticBeanstalk', { minInputs: 1, maxInputs: 1, minOutputs: 0, maxOutputs: 0 }, {
+//             ApplicationName: applicationName,
+//             EnvironmentName: environmentName
+//         });
+//     }
+// }
+
+// export class OpsWorksDeploy extends DeployAction {
+//     constructor(parent: Stage, name: string, app: string, stack: string, layer?: string) {
+//         super(parent, name, 'OpsWorks', { minInputs: 1, maxInputs: 1, minOutputs: 0, maxOutputs: 0 }, {
+//             Stack: stack,
+//             App: app,
+//             Layer: layer,
+//         });
+//     }
+// }
+
+// export class ECSDeploy extends DeployAction {
+//     constructor(parent: Stage, name: string, clusterName: string, serviceName: string, fileName?: string) {
+//         super(parent, name, 'ECS', { minInputs: 1, maxInputs: 1, minOutputs: 0, maxOutputs: 0 }, {
```

Comment:
Can we save this code somewhere else (issue?) instead of commented out.. it’s a bit messy

Reply:
I was referring to the commented out code. Let's remove it from here. If you want to "save" it somewhere, just paste into your issue.